### ISSUE TITLE
Allow users to enable non-root auth in simulation

### ISSUE THREAD

**Issue by dmkozh** [stellar/stellar-rpc#344]
### What problem does your feature solve?

Currently simulation by default disallows 'non-root' authorization in recording mode, i.e. disallows potentially 'frontrunnable' authorizations. This is a very sensible default setting as most of the time it's correct to make all the calls non-frontrunnable. That said, there is still a range of cases where non-root auth is intentional and thus we need to provide users with a way of enabling this.

### What would you like to see?

There should be some request parameter that enables non-root auth. It's fine for this to be not easily discoverable or accessible as this is meant for the advanced users. But there needs to be at least _some_ way of enabling this.

### What alternatives are there?

Not providing the users with an option. This basically makes auth simulation unusable for a fraction of the users, which is not desirable.

**Comment by dmkozh** [comment]
This requires an env release as well. I'll work on this and assign this to myself for the time being. It would be great if someone else could help with the RPC interface part when the env change is released.

**Comment by dmkozh** [comment]
The env-side change has been released in 22.1.3 release (https://github.com/stellar/rs-soroban-env/pull/1503) and it can be now implemented in soroban-rpc at any time now. 

It had to be guarded by `unstable-next-api` feature in order to pass the semver checks, which should be fine to enable in soroban-rpc as we'll be aware of the dependency and will make sure to not accidentally break it (should there be any more API-breaking features before v23 release).

The alternative is to wait until the stable v23 release, though it might be a good idea to at least prepare the change before that.

**Comment by Shaptic** [comment]
@leighmcculloch @dmkozh is there a way to programmatically determine whether or not non-root authorization is present from the authorization entries themselves? My gut says "no" but I just want to make sure - it would help us avoid introducing + propagating an additional request parameter. My gut also says "this should be opt-in" since per Dima this only applies to a small subset of simulations.

Otherwise, I'm proposing the following:

```typescript
interface SimulateTransactionRequest { 
   // ...
  authMode?: "enforce" | "record" | "nonroot"
}
```

where

* **omission** (default): enforcement if auth entries are included in the invocation operation, root recording otherwise (the behavior today)
* **enforce**: enforcement always, with an empty list if auth entries are omitted
* **record**: root record always, ignoring the list if auth entries are included
* **nonroot**: record with non-root authorization, ignoring the list if auth entries are included (naming? could be `record-nonroot`)

Thoughts? Is this overkill? A simpler `nonroot?: boolean` would be a direct parallel to the simulation library.

**Comment by Shaptic** [comment]
>  The default behavior becomes a bit muddy though as the default mode can not be expressed with authMode (because it's based on presence of the auth entries).

True, but this is the way it works now (entries = enforcing, no entries = recording) which means it's non-breaking behavior. For your scenario (partial recording), we could instead do one of three things:
1. Today, `authMode: "record"` fails if auth entries are present. Tomorrow, it does partial recording.
2. Today, `authMode: "record"` ignores auth entries. Tomorrow, we have breaking behavior where it will do partial recording.
3. Today and tomorrow, `authMode: "record"` ignores auth entries and we add a `authMode: "partial_record"` that doesn't ignore them.

I'm partial to (1): it does seem like erratic behavior if you include entries and want recording mode. In the custom auth case today, the second simulation uses enforcing mode.

> if we could use a bit longer name than non-root, that would be great

👍 sounds good

**Comment by dmkozh** [comment]
>  we could instead do one of three things:

Sure, my main point is that `authMode` seems more scalable and flexible than just a flag.
